File: python/ex3.py
import math

epsilon = 0.000001
limit = 1000
max_denominator = 10000

def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

def is_possibly_rational(x):
    for b in range(1, max_denominator + 1):
        a = round(x * b)
        if abs(x - a / b) < epsilon:
            common_divisor = gcd(a, b)
            print(f"{a // common_divisor}/{b // common_divisor}")
            return 1
    return -1

def print_result(num):
    if num == -1:
        print("infinity")
    elif is_possibly_rational(num) == 1:
        pass
    else:
        print("irrational")

def input_values():
    a, b = map(int, input().split())
    return a, b

def summing(a, b):
    cur = 1 / b
    sum_past = cur
    sum_value = cur

    for n in range(2, limit):
        cur = (n ** a) / (b ** n)
        sum_value += cur
        if abs(sum_value - sum_past) < epsilon:
            return sum_value
        sum_past = sum_value
    return -1

def main():
    a, b = input_values()
    res = summing(a, b)
    print_result(res)

if __name__ == "__main__":
    main()

File: c++/ex3.cpp
#include <iostream>
#include <cmath>


#define epsilon                0.000001
#define limit                  1000
#define max_denominator        10000

int gcd(int a, int b) {
    return (b == 0) ? a : gcd(b, a % b);
}

int is_possibly_rational(double x) {
    for (int b = 1; b <= max_denominator; b++) {
        int a = (int)round(x * b);
        if (fabs(x - (double)a / b) < epsilon) {
            int common_divisor = gcd(a, b);
            std::cout << a / common_divisor << "/" <<  b / common_divisor << std::endl;
            return 1;
        }
    }
    return -1;
}

void print(double num){
    if (num == -1){
        std::cout << "infinity" << std::endl;
    }
    else if (is_possibly_rational(num) == 1){
    }
    else{
        std::cout << "irrational"<< std::endl;
    }
}

void input(int& a,int& b){
    std::cin >> a >> b;
}

double summing(int a,int b){
    double cur;
    double sum_past = 0,sum = 0;

    cur = 1./b;
    sum_past = cur;
    sum = cur;

    for (int n = 2;n<limit;n ++){
        cur = pow(n,a)/pow(b,n);
        sum += cur;
        if (fabs(sum-sum_past) < epsilon){
            return sum;
        }
        sum_past = sum;
    }
    return -1;
}

int main(void){
    int a,b;

    input(a,b);

    double res = summing(a,b);

    print(res);
}